### PR TITLE
[AERIE-1709]  Avoid regenerating activities after scheduling

### PR DESCRIPTION
**Pull request by adrienmaillard**
* **Tickets addressed:** AERIE-1709
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

This ticket has been created to address the following behavior of the scheduler server: when finished with a scheduling request, instead of sending updates to merlin, the plan was deleted and then regenerated from scratch. This updated the ids of activities even though their properties were not affected. 

The approach taken has been to discriminate between activities to add, activities to delete, and activities to update.
For that, we need to compare the initial plan with the new plan. To do that, we introduce two new classes representing simple merlin activities and merlin plans.  

When fetching the initial plan before scheduling, we save a version of the merlin distant plan for later comparison. We also save a mapping from merlin activity instance ids to scheduler activity instance id. 

Then after scheduling, we loop through activities from the new plan:  
If an activity with the same id is not present in the old plan, it is a new activity, we insert it in the distant merlin plan. 
If it is present, we check if it has changed by comparing their two merlin representations and if they are not the same, we update the merlin instance with the one from the plan (the most "recent"). 
And if an activity is present in the old plan but absent from the new plan, it means that the activity was deleted during scheduling, we then delete it from the distant merlin plan. 

The first three commits are performing these changes. 

The next four commits are removing some useless code and organizing the scheduler into packages. 

## Verification

For changes not related with the ticket: all tests are passing.  

For changes related with the ticket, I performed tests from the UI with the bananation mission model. When hitting several times the scheduler, even if new activities are inserted in the plan, the ids of activities already present do not change. 
 
ids

## Documentation

Please check the validity of the update/delete/add model.

## Future work

None regarding this issue.

### CODE REVIEW

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Plan.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Plan.java]: 🔻 In keeping with the rest of the Aerie code base, we can add these name spaces as explicit imports and remove them here and anywhere else added in this PR.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/MerlinPlan.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/MerlinPlan.java]: :large_orange_diamond: Should these parameters be `final`?

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/MerlinPlan.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/MerlinPlan.java]: 🔶 Should this parameter be final?

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]: :large_orange_diamond: Let's use multi-line string here e.g. 
`"some string text"`
`+ "more string"`
`+ "more"`

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/MerlinActivityInstance.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/MerlinActivityInstance.java:1]: :large_orange_diamond: this file has a number of extra new lines that can be removed and each parameter in the constructor prototype can take a new line.

File: scheduler-server/src/test/java/gov/nasa/jpl/aerie/scheduler/server/services/SchedulingDSLCompilationServiceTests.java
Comment (JoelCourtney) [inline scheduler-server/src/test/java/gov/nasa/jpl/aerie/scheduler/server/services/SchedulingDSLCompilationServiceTests.java]: (not a review) What's the difference between a `Plan` and a `MerlinPlan`?

Comment (adrienmaillard) [reply scheduler-server/src/test/java/gov/nasa/jpl/aerie/scheduler/server/services/SchedulingDSLCompilationServiceTests.java]: A `Plan` is the scheduler's representation of a plan, with partially instanciated activity parameters, activity durations and so on.  A `MerlinPlan` is limited to the representation of a plan in the merlin database.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (mattdailis) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]: Just in terms of organization of the code - it seems a bit out of place for `GraphQLMerlinService` to know about `Problem`. (I know this was the case before this PR, just noticing this now).

So I guess I would expect MerlinService to return a `MerlinPlan`, and then `loadInitialPlan` can convert a `MerlinPlan` into a `Plan`.

Comment (adrienmaillard) [reply scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]: Sounds cleaner, I am going to do that.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (mattdailis) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]: IntelliJ is telling me that we can use `entrySet()` instead of `keySet()`, so we won't need to call `.get()` below.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java
Comment (mattdailis) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java]: I know we only use this `Triple` in one place, but I think it would help readability (on the caller's side) to define a record:

```
record PlanComponents(Plan schedulerPlan, MerlinPlan merlinPlan, Map<SchedulingActivityInstanceId, ActivityInstanceId> idMap) {}
```
It can go right below this function, next to the `SchedulerMissionModel` record, which serves a similar purpose.

This way, the caller code can refer to `loadedPlanComponents.idMap()` and  `loadedPlanComponents.merlinPlan()` instead of `loadedPlanComponents.getRight()` and  `loadedPlanComponents.getMiddle()`

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Plan.java
Comment (patkenneally) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Plan.java]: 🔻 These Java docs seem like they were copy and pasted but not edited for their new context.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]: :large_orange_diamond: `final var` for these?

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]: :large_orange_diamond: `final` ?

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]: :large_blue_circle: this reads easier with `.parse` dropped onto a new line.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]: :large_blue_circle: Some extra new lines here.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]: :large_orange_diamond: I'm unclear about why this and above are throwing a `NoSuchPlanException`, and with a null reason. Can you help me understand your reasoning?

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]: :large_orange_diamond: Extra new line

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]: :small_red_triangle_down: Should these be `id.id()` and `planId.id()` ? Also for the variable name `id` it seems like it would read better as something like `instanceId` or similar.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java:385]: :small_red_triangle_down: The same question as above; should this truly be a `NoSuchPlanException` ? Or is something more akin to some parsing exception appropriate?

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java]: :large_orange_diamond: We lost the `final` here. I'd also be partial to dropping the method parameters each to a new line so this line doesn't run so far off to the right.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java
Comment (patkenneally) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java]: :large_orange_diamond: `final` ?

**Review by patkenneally — CHANGES_REQUESTED** [review]
:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired

Mostly formatting changes :)

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (Twisol) [reply scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]:
```java
"""
These work too, and were added a
few versions of Java
ago.
"""
```

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (adrienmaillard) [reply scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java]: In general, this exception as well as `NoSuchMissionModelException` are thrown because we **assume** this is the reason it fails. In this case, it does not make a lot of sense so I am going to replace it with a new exception `NoSuchActivityInstanceException`.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java
Comment (adrienmaillard) [reply scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java:385]: See my above comment about the other case. In this case, there are more reasons why it may fail (plan is absent or activity is absent) so we pick one. Failure at the class cast or arithmetic level only means that something is wrong with the request. In other words, we assume the json unpacking is right on our side and they are proxy exception for the real issue.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Plan.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Plan.java]: Did this in `Plan` and `PlanInMemory`.

**Comment by adrienmaillard** [comment]
I've just force-pushed changes to address all of the above comments.

**Review by patkenneally — APPROVED** [review]

**Review by mattdailis — APPROVED** [review]
Looks good to me, thanks for making those changes 💯

**Pull request merged**
